### PR TITLE
[WIP] Multiple filters support

### PR DESCRIPTION
**Pull request by bbotella**
We still lack composition. With this approach, we define some filters, and we can add a composition string to the options that would go like:

`(filter_name1 AND filter_name2) OR filter_name3`

Maybe we may need a custom interpreter to parse this composition string. Thoughts @eliasdorneles?

### CODE REVIEW

File: tests/test_readers_s3.py [hunk 1]
```diff
@@ -270,6 +272,7 @@ def test_get_batch(self):
         self.assertEqual(batch, expected_batch)
         shutil.rmtree(reader.tmp_folder, ignore_errors=True)
 
+    @freeze_time('2010-01-01')
```
Comment (bbotella) [inline tests/test_readers_s3.py]: I added this as some tests failed locally due to a time change. Maybe I should do another PR for it to master branch directly.

File: exporters/filters/multiple_filter.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        self.filters = self.load_filters()
+        self.composition = self.read_option('composition')
+
+    def load_filters(self):
+        filters = {}
+        for filter_name, filter_config in self.read_option('filters').iteritems():
+            filters[filter_name] = self.module_loader.load_filter(filter_config, self.metadata)
+        return filters
+
+    def and_filter(self, item):
```
Comment (bbotella) [inline exporters/filters/multiple_filter.py]: `intersection_filter` may be a better name.

**Comment by eliasdorneles** [comment]
Hm, perhaps it would be nice to do some designing on the user interface first. :)
I mean, how are we expecting people to use this?

It would be nice to arrive at something more or less usable (and preferably something that allowed people to copy'n paste filters).

So, here is one idea, using lists to represent the operations:

```
"filter": {
    "name": "exporters.filters.MultipleFilter",
    "options": {
        "filters": ["or",
            {"name": "exporters.filters.PythonexpFilter", "options": {...}},
            {"name": "exporters.filters.KeyValueFilter", "options": {...}},
            ["and",
                {"name": "exporters.filters.PythonexpFilter", "options": {...}},
                {"name": "exporters.filters.KeyValueFilter", "options": {...}},
            ]
        ]
    }
}
```

Here is another, using dicts:

```
"filter": {
    "name": "exporters.filters.MultipleFilter",
    "options": {
        "filters": {
          "or": [
            {"name": "exporters.filters.PythonexpFilter", "options": {...}},
            {"name": "exporters.filters.KeyValueFilter", "options": {...}},
            {"and": [
                {"name": "exporters.filters.PythonexpFilter", "options": {...}},
                {"name": "exporters.filters.KeyValueFilter", "options": {...}},
            ]},
            {"name": "exporters.filters.KeyValueRegexFilter", "options": {...}},
        ]}
    }
}
```

I like more the syntax of this last one, but it will require more validation than the latter (because it can't allow a dict representing a filter to have both "and" and "or" in the same object).

**Comment by eliasdorneles** [comment]
I see forgot to ask... what do you think? :)

**Comment by bbotella** [comment]
My idea was to support something like:

```
{
            'filters': {
                'filter1': {
                    'name': 'exporters.filters.KeyValueFilter',
                    'options': {
                        'keys': [{'name': 'country_code', 'value': 'es', 'operator': 'contains'}]
                    }
                },
                'filter2': {
                    'name': 'exporters.filters.KeyValueFilter',
                    'options': {
                        'keys': [{'name': 'name', 'value': 'item1', 'operator': 'contains'}]
                    }
                },
                'filter3': {
                    'name': 'exporters.filters.KeyValueFilter',
                    'options': {
                        'keys': [{'name': 'name', 'value': 'item3', 'operator': 'contains'}]
                    }
                }
            }
        },
        'composition': '(filter1 and filter2) or filter3'
```

This would allow us to make clear and user friendly compositions, only by replacing filter names in composition string by their filter value, and running that resulting composition string on the python interpreter (that would be something like '(True and False) or True')

**Comment by eliasdorneles** [comment]
Hmm, as an user, I would not like all that indirection (having to come up with names for filters and having to type them twice) and it doesn't look like this would make implementation simpler.

I also don't think that having to do string replacement for _every_ record passing over the filter and evaluating it is an advantage for that approach.

With any of the two options I mentioned, the filter would be "compiled" into a function in the constructor (Python code for the "and" and "or") and then calling the filter for each record would be simply that function call (no need for eval).

**Comment by eliasdorneles** [comment]
Btw, a good source of inspiration is [MongoDB query filters](https://docs.mongodb.com/manual/tutorial/query-documents/#read-operations-query-argument). :)

**Comment by tsrdatatech** [comment]
@bbotella @eliasdorneles we have this PR https://github.com/scrapinghub/exporters/pull/325 that is almost done, do you think it makes sense to keep this one or close it?

**Comment by eliasdorneles** [comment]
We can close it, yes, this was another PR done as an experiment and discussion starter. =)
Thanks @bbotella ! <3
